### PR TITLE
PERF: regression in DataFrame reduction ops performance #37081

### LINKED ISSUE #37081: PERF: regression in DataFrame reduction ops performance
From https://pandas.pydata.org/speed/pandas/#stat_ops.FrameOps.time_op?Cython=0.29.21&Cython=0.29.16&p-op='sum'&p-dtype='int'&p-axis=0&commits=3a043f2d-4c03d07b&x-axis-scale=date

Reproducer:

```
values = np.random.randn(100000, 4)  
df = pd.DataFrame(values).astype("int")
%timeit df.sum()
```

increased with a  factor 2 to 3 x somewhere the last days.

### PR DESCRIPTION
**Pull request by ukarroum**
- [x] closes #37081 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Made the change proposed by @jorisvandenbossche in https://github.com/pandas-dev/pandas/pull/35881#discussion_r504035006 

Did a very quick comparison : 

With `self.dtypes` (old version) : 

```
In [8]: values = np.random.randn(100000, 4)   
   ...: df = pd.DataFrame(values).astype("int") 
   ...: %timeit df.sum() 
714 µs ± 21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

With `self._iter_column_arrays()` (new version) : 

```
In [4]: values = np.random.randn(100000, 4)   
   ...: df = pd.DataFrame(values).astype("int") 
   ...: %timeit df.sum() 
477 µs ± 8.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
do we have an asv that covers this case, if not can you add one?

this doesn't need a note as its on master.

**Comment by ukarroum** [comment]
> do we have an asv that covers this case, if not can you add one?
> 
> this doesn't need a note as its on master.

I believe we do : 

https://pandas.pydata.org/speed/pandas/#stat_ops.FrameOps.time_op?p-op='sum'&p-dtype='int'

**Comment by jorisvandenbossche** [comment]
Indeed. Jeff, see the issue, it was actually catched thanks to our asv suite

File: pandas/core/frame.py [hunk 1]
```diff
@@ -8670,7 +8670,10 @@ def _reduce(
             cols = self.columns[~dtype_is_dt]
             self = self[cols]
 
-        any_object = self.dtypes.apply(is_object_dtype).any()
+        any_object = np.array(
+            [is_object_dtype(values.dtype) for values in self._iter_column_arrays()],
+            dtype=bool,
+        ).any()
```
Comment (jbrockmendel) [inline pandas/core/frame.py:8676]: let's only find the dtypes once (i.e. share with dtype_is_dt above

```
own_dtypes = [arr.dtype for arr in self._iter_column_arrays()]
# or 
own_dtypes = [blk.dtype for blk in self._mgr.blocks]
```

Comment (ukarroum) [reply pandas/core/frame.py:8675]: done in 41827fb

**Review by jorisvandenbossche — APPROVED** [review]
Thanks @ukarroum, looks good!

**Comment by ukarroum** [comment]
Should i do something about the : 2 failed azure pipelines ?
Looks like they're failing on master too and the `./ci/code_checks.sh` localy return no error.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ukarroum

**Pull request merged**